### PR TITLE
Create a new AIOCallback to call back aio_err_callbck if there is an error from Linux Native AIO

### PR DESCRIPTION
**Pull request by oknet**

### CODE REVIEW

**Comment by oknet** [comment]
This is a bugfix for the aio sub-system with Linux Native AIO enabled.

It is necessory to create a new AIOCallback to call back the ```aio_err_callbck```.

File: iocore/aio/AIO.cc [hunk 1]
```diff
@@ -587,7 +576,13 @@ DiskHandler::mainAIOEvent(int event, Event *e)
   }
 
   while ((op = complete_list.dequeue()) != nullptr) {
-    op->handleEvent(event, e);
+    op->mutex = op->action.mutex;
+    MUTEX_TRY_LOCK(lock, op->mutex, trigger_event->ethread);
+    if (!lock.is_locked()) {
+      trigger_event->ethread->schedule_imm(op);
```
Comment (zizhong) [inline iocore/aio/AIO.cc]: some delay here is better?

File: iocore/aio/P_AIO.h [hunk 2]
```diff
@@ -130,6 +105,25 @@ struct AIO_Reqs {
...
+  (void)data;
+  if (aio_err_callbck && !ok()) {
+    AIOCallback *err_op          = new AIOCallbackInternal();
+    err_op->aiocb.aio_fildes     = this->aiocb.aio_fildes;
+    err_op->aiocb.aio_lio_opcode = this->aiocb.aio_lio_opcode;
+    err_op->mutex                = aio_err_callbck->mutex;
+    err_op->action               = aio_err_callbck;
+    eventProcessor.schedule_imm(err_op);
+  }
+  if (!action.cancelled)
```
Comment (zizhong) [inline iocore/aio/P_AIO.h]: You are combining the to io_complete() together. In the original native AIO one, it has the process to acquire a mutex.  Do we need it here?

File: iocore/aio/AIO.cc (hunk 1 above)
Comment (oknet) [reply iocore/aio/AIO.cc]: Yes, it makes the DiskHandler running in non-blocked. 
In the original native AIO's `io_complete`, it always calls `SCOPE_MUTEX_LOCK` to acquire lock in blocked way.

File: iocore/aio/P_AIO.h (hunk 2 above)
Comment (oknet) [reply iocore/aio/P_AIO.h]: It does with `DiskHandler::mainAIOEvent`.
To call back any `handleEvent()`, It is required to get the mutex locked.

Comment (zizhong) [reply iocore/aio/P_AIO.h]: I see. Thanks!

File: iocore/aio/AIO.cc [hunk 3]
```diff
@@ -514,9 +505,7 @@ aio_thread_main(void *arg)
       op->mutex     = op->action.mutex;
       if (op->thread == AIO_CALLBACK_THREAD_AIO) {
         SCOPED_MUTEX_LOCK(lock, op->mutex, thr_info->mutex->thread_holding);
-        if (!op->action.cancelled) {
-          op->action.continuation->handleEvent(AIO_EVENT_DONE, op);
-        }
+        op->handleEvent(EVENT_NONE, nullptr);
```
Comment (scw00) [inline iocore/aio/AIO.cc]: Why do you remove this check ?

File: iocore/aio/AIO.cc [hunk 4]
```diff
@@ -587,7 +576,13 @@ DiskHandler::mainAIOEvent(int event, Event *e)
   }
 
   while ((op = complete_list.dequeue()) != nullptr) {
-    op->handleEvent(event, e);
+    op->mutex = op->action.mutex;
```
Comment (scw00) [inline iocore/aio/AIO.cc]: I think aio should consider the user's lock.

File: iocore/aio/AIO.cc (hunk 3 above)
Comment (scw00) [reply iocore/aio/AIO.cc]: Drop it ! You callback to io_complete instead.

Comment (oknet) [reply iocore/aio/AIO.cc]: It will be checked within `io_complete`.

File: iocore/aio/AIO.cc (hunk 4 above)
Comment (oknet) [reply iocore/aio/AIO.cc]: I'm moving it into `io_complete`.

Comment (scw00) [reply iocore/aio/AIO.cc]: Yes you are right ! the original code locked the CacheVC.

**Comment by zwoop** [comment]
Is this a candidate for 7.1.x ?

**Comment by oknet** [comment]
@zwoop 

It is not necessary for 7.1.x, because:

- The bug affects only the Linux Native AIO mode and the mode is not enabled by default
- The bug is only triggered when a disk error occurs

**Pull request merged**